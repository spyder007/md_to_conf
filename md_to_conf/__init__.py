#!/usr/bin/env python3
import logging
import sys
import os
import argparse
from .confluence_converter import ConfluenceConverter
from .client import ConfluenceApiClient, PageInfo
from .converter import MarkdownConverter


def get_parser():
    PARSER = argparse.ArgumentParser()
    PARSER.add_argument(
        "markdownFile", help="Full path of the markdown file to convert and upload."
    )
    PARSER.add_argument(
        "spacekey",
        help="Confluence Space key for the page. If omitted, will use user space.",
    )
    PARSER.add_argument(
        "-u", "--username", help="Confluence username if $CONFLUENCE_USERNAME not set."
    )
    PARSER.add_argument(
        "-p", "--apikey", help="Confluence API key if $CONFLUENCE_API_KEY not set."
    )
    PARSER.add_argument(
        "-o",
        "--orgname",
        help="Confluence organisation if $CONFLUENCE_ORGNAME not set. "
        "e.g. https://XXX.atlassian.net/wiki"
        "If orgname contains a dot, considered as the fully qualified domain name."
        "e.g. https://XXX",
    )
    PARSER.add_argument(
        "-a",
        "--ancestor",
        help="Parent page under which page will be created or moved.",
    )
    PARSER.add_argument(
        "-t",
        "--attachment",
        nargs="+",
        help="Attachment(s) to upload to page. Paths relative to the markdown file.",
    )
    PARSER.add_argument(
        "-c",
        "--contents",
        action="store_true",
        default=False,
        help="Use this option to generate a contents page.",
    )
    PARSER.add_argument(
        "-n",
        "--nossl",
        action="store_true",
        default=False,
        help="Use this option if NOT using SSL. Will use HTTP instead of HTTPS.",
    )
    PARSER.add_argument(
        "-d",
        "--delete",
        action="store_true",
        default=False,
        help="Use this option to delete the page instead of create it.",
    )
    PARSER.add_argument(
        "-l",
        "--loglevel",
        default="INFO",
        help="Use this option to set the log verbosity.",
    )
    PARSER.add_argument(
        "-s",
        "--simulate",
        action="store_true",
        default=False,
        help="Use this option to only show conversion result.",
    )
    PARSER.add_argument(
        "-v",
        "--version",
        type=int,
        action="store",
        default=2,
        help="Version of confluence page (default is 2).",
    )
    PARSER.add_argument(
        "-mds",
        "--markdownsrc",
        action="store",
        default="default",
        choices=["default", "bitbucket"],
        help="Use this option to specify a markdown source "
        " (i.e. what processor this markdown was targeting). "
        "Possible values: bitbucket.",
    )
    PARSER.add_argument(
        "--label",
        action="append",
        dest="labels",
        default=[],
        help="A list of labels to set on the page.",
    )
    PARSER.add_argument(
        "--property",
        action="append",
        dest="properties",
        default=[],
        type=lambda kv: kv.split("="),
        help="A list of content properties to set on the page.",
    )
    PARSER.add_argument(
        "--title",
        action="store",
        dest="title",
        default=None,
        help="Set the title for the page, otherwise the title is "
        "going to be the first line in the markdown file",
    )
    PARSER.add_argument(
        "--remove-emojies",
        action="store_true",
        dest="remove_emojies",
        default=False,
        help="Remove emojies if there are any. This may be need if "
        "the database doesn't support emojies",
    )

    return PARSER


def validate_args(user_name, api_key, markdown_file, org_name):
    LOGGER = logging.getLogger(__name__)
    if user_name is None:
        LOGGER.error("Error: Username not specified by environment variable or option.")
        sys.exit(1)

    if api_key is None:
        LOGGER.error("Error: API key not specified by environment variable or option.")
        sys.exit(1)

    if not os.path.exists(markdown_file):
        LOGGER.error("Error: Markdown file: %s does not exist.", markdown_file)
        sys.exit(1)

    if org_name is None:
        LOGGER.error("Error: Org Name not specified by environment variable or option.")
        sys.exit(1)


def main():
    """
    Main program

    """
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s - \
        %(levelname)s - %(funcName)s [%(lineno)d] - \
        \t%(message)s",
    )
    LOGGER = logging.getLogger(__name__)

    # ArgumentParser to parse arguments and options
    PARSER = get_parser()

    ARGS = PARSER.parse_args()

    # Assign global variables
    try:
        # Set log level
        LOGGER.setLevel(getattr(logging, ARGS.loglevel.upper(), None))

        MARKDOWN_FILE = ARGS.markdownFile
        SPACE_KEY = ARGS.spacekey
        USERNAME = os.getenv("CONFLUENCE_USERNAME", ARGS.username)
        API_KEY = os.getenv("CONFLUENCE_API_KEY", ARGS.apikey)
        ORGNAME = os.getenv("CONFLUENCE_ORGNAME", ARGS.orgname)
        ANCESTOR = ARGS.ancestor
        NOSSL = ARGS.nossl
        DELETE = ARGS.delete
        SIMULATE = ARGS.simulate
        VERSION = ARGS.version
        MARKDOWN_SOURCE = ARGS.markdownsrc
        LABELS = ARGS.labels
        PROPERTIES = dict(ARGS.properties)
        ATTACHMENTS = ARGS.attachment
        CONTENTS = ARGS.contents
        TITLE = ARGS.title
        REMOVE_EMOJIES = ARGS.remove_emojies

        validate_args(USERNAME, API_KEY, MARKDOWN_FILE, ORGNAME)

    except Exception as err:
        LOGGER.error("\n\nException caught:\n%s ", err)
        LOGGER.error("\nFailed to process command line arguments. Exiting.")
        sys.exit(1)

    LOGGER.info("\t----------------------------------")
    LOGGER.info("\tMarkdown to Confluence Upload Tool")
    LOGGER.info("\t----------------------------------")

    LOGGER.info("Markdown file:\t%s", MARKDOWN_FILE)
    LOGGER.info("Space Key:\t%s", SPACE_KEY)

    confluence_converter: ConfluenceConverter = ConfluenceConverter(
        MARKDOWN_FILE,
        MARKDOWN_SOURCE,
        TITLE,
        ORGNAME,
        not NOSSL,
        USERNAME,
        SPACE_KEY,
        API_KEY,
        ANCESTOR,
        VERSION,
    )

    if DELETE:
        confluence_converter.delete_page(SIMULATE)
        return

    confluence_converter.convert(
        SIMULATE, REMOVE_EMOJIES, CONTENTS, LABELS, PROPERTIES, ATTACHMENTS
    )
