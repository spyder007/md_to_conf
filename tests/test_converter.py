import pytest
from md_to_conf import MarkdownConverter


URL = "https://domain.confluence.net/wiki"


@pytest.fixture
def test_converter_basic() -> MarkdownConverter:
    return MarkdownConverter("tests/testfiles/basic.md", URL, "default", 2)


@pytest.fixture
def test_converter_advanced() -> MarkdownConverter:
    return MarkdownConverter("tests/testfiles/advanced.md", URL, "default", 2)


@pytest.fixture
def test_converter_image() -> MarkdownConverter:
    return MarkdownConverter("tests/testfiles/withimage.md", URL, "default", 2)


def test_converter_init():
    md_file = "tests/testfiles/basic.md"
    url = "https://domain.confluence.net/wiki"
    md_source = "default"
    editor_version = 2

    converter = MarkdownConverter(md_file, url, md_source, editor_version)

    assert converter.md_file == md_file
    assert converter.api_url == url
    assert converter.md_source == md_source
    assert converter.editor_version == editor_version


def test_converter_basic_test(test_converter_basic: MarkdownConverter, snapshot):
    html = test_converter_basic.get_html_from_markdown()
    assert html == snapshot


def test_converter_image_test(test_converter_image: MarkdownConverter, snapshot):
    html = test_converter_image.get_html_from_markdown()
    assert html == snapshot


def test_converter_html(test_converter_basic: MarkdownConverter, snapshot):
    html = test_converter_basic.convert_md_to_conf_html(
        remove_first_line=True, remove_emojies=False, add_contents=False
    )
    assert html == snapshot


def test_converter_html_advanced_with_toc(
    test_converter_advanced: MarkdownConverter, snapshot
):
    html = test_converter_advanced.convert_md_to_conf_html(
        remove_first_line=True, remove_emojies=True, add_contents=True
    )
    assert html == snapshot


def test_converter_html_advanced(test_converter_advanced: MarkdownConverter, snapshot):
    html = test_converter_advanced.convert_md_to_conf_html(
        remove_first_line=True, remove_emojies=True, add_contents=False
    )
    assert html == snapshot


def test_slug(test_converter_basic: MarkdownConverter):
    slug = test_converter_basic.slug("<tag>The $slug</tag>", False)
    assert slug == "The-slug"


def test_slug_lower(test_converter_basic: MarkdownConverter):
    slug = test_converter_basic.slug("<tag>The $slug</tag>", True)
    assert slug == "the-slug"


def test_convert_comment_block(test_converter_basic: MarkdownConverter):
    slug = test_converter_basic.convert_comment_block("<!-- some comments go here -->")
    assert slug == "<ac:placeholder> some comments go here </ac:placeholder>"


def test_toc_convert(test_converter_basic: MarkdownConverter, snapshot):
    slug = test_converter_basic.create_table_of_content(
        "<h1>First heading</h1> <p>[TOC]</p> <h2>Second heading</h2>"
    )

    assert slug == snapshot
