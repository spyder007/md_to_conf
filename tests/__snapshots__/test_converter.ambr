# serializer version: 1
# name: test_converter_basic_test
  '''
  <h1>Basic Page</h1>
  <h2>Heading 2</h2>
  <p>Some text for heading 2</p>
  <h3>Heading 3-1</h3>
  <p>Some text for heading 3-1</p>
  <h3>Heading 3-2</h3>
  <p>Some test for heading 3-2</p>
  <h4>Unordered list</h4>
  <ul>
  <li>An</li>
  <li>unordered</li>
  <li>list</li>
  </ul>
  <h4>Ordered List</h4>
  <ol>
  <li>An</li>
  <li>Ordered</li>
  <li>List</li>
  </ol>
  <h4>Info Block</h4>
  <blockquote>
  <p>Info block</p>
  </blockquote>
  <h4>Note block</h4>
  <blockquote>
  <p>Note: This is a note</p>
  </blockquote>
  <h4>Warning Block</h4>
  <blockquote>
  <p>Warning: This is a warning</p>
  </blockquote>
  '''
# ---
# name: test_converter_html
  '''
  <h2>Heading 2</h2>
  <p>Some text for heading 2</p>
  <h3>Heading 3-1</h3>
  <p>Some text for heading 3-1</p>
  <h3>Heading 3-2</h3>
  <p>Some test for heading 3-2</p>
  <h4>Unordered list</h4>
  <ul>
  <li>An</li>
  <li>unordered</li>
  <li>list</li>
  </ul>
  <h4>Ordered List</h4>
  <ol>
  <li>An</li>
  <li>Ordered</li>
  <li>List</li>
  </ol>
  <h4>Info Block</h4>
  <p><ac:structured-macro ac:name="info"><ac:rich-text-body><p>Info block</p></ac:rich-text-body></ac:structured-macro></p>
  <h4>Note block</h4>
  <p><ac:structured-macro ac:name="note"><ac:rich-text-body><p>This is a note</p></ac:rich-text-body></ac:structured-macro></p>
  <h4>Warning Block</h4>
  <p><ac:structured-macro ac:name="warning"><ac:rich-text-body><p>This is a warning</p></ac:rich-text-body></ac:structured-macro></p>
  '''
# ---
# name: test_converter_html_advanced
  '''
  <h2>Headings</h2>
  <p>The heading for this section is a level 2 heading</p>
  <h3>Level 3</h3>
  <p>Some text for level 3</p>
  <h4>Level 4</h4>
  <p>Some text for level 4</p>
  <h5>Level 5</h5>
  <p>Some text for level 5</p>
  <h2>Lists</h2>
  <h3>Unordered list</h3>
  <ul>
  <li>An</li>
  <li>unordered</li>
  <li>list</li>
  </ul>
  <h3>Ordered list</h3>
  <ol>
  <li>An</li>
  <li>Ordered</li>
  <li>List</li>
  </ol>
  <h3>Nested Unordered list</h3>
  <ul>
  <li>This list is nested</li>
  <li>This item has children<ul>
  <li>This is a child of item 2</li>
  <li>Another child of item 2</li>
  </ul>
  </li>
  <li>This is the third item</li>
  </ul>
  <h3>Nested Ordered list</h3>
  <ol>
  <li>This list is nested</li>
  <li>This item has children<ol>
  <li>This is a child of item 2</li>
  <li>Another child of item 2</li>
  </ol>
  </li>
  <li>This is the third item</li>
  </ol>
  <h2>Blocks</h2>
  <h3>Info Block</h3>
  <p><ac:structured-macro ac:name="info"><ac:rich-text-body><p>This should be an info block</p></ac:rich-text-body></ac:structured-macro></p>
  <h3>Note/Warning Block</h3>
  <p><ac:structured-macro ac:name="note"><ac:rich-text-body><p>This should be a note block</p></ac:rich-text-body></ac:structured-macro></p>
  <h3>Warning Block</h3>
  <p><ac:structured-macro ac:name="warning"><ac:rich-text-body><p>This should be a warning block</p></ac:rich-text-body></ac:structured-macro></p>
  <h2>Code</h2>
  <h3>No language</h3>
  <ac:structured-macro ac:name="code"><ac:parameter ac:name="theme">Midnight</ac:parameter><ac:parameter ac:name="linenumbers">true</ac:parameter><ac:parameter ac:name="language">none</ac:parameter><ac:plain-text-body><![CDATA[This is a code block with on language assigned
  ]]></ac:plain-text-body></ac:structured-macro>
  <h3>Javascript</h3>
  <ac:structured-macro ac:name="code"><ac:parameter ac:name="theme">Midnight</ac:parameter><ac:parameter ac:name="linenumbers">true</ac:parameter><ac:parameter ac:name="language">javascript</ac:parameter><ac:plain-text-body><![CDATA[var block = has_javascript_lang();
  ]]></ac:plain-text-body></ac:structured-macro>
  <h2>Tables</h2>
  <table>
  <thead>
  <tr>
  <th>Heading 1</th>
  <th>Heading 2</th>
  <th>Heading 3</th>
  </tr>
  </thead>
  <tbody>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  </tbody>
  </table>
  '''
# ---
# name: test_converter_html_advanced_with_toc
  '''
  <ac:structured-macro ac:name="toc">
  <ac:parameter ac:name="printable">true</ac:parameter>
  <ac:parameter ac:name="style">disc</ac:parameter><ac:parameter ac:name="maxLevel">5</ac:parameter>
  <ac:parameter ac:name="minLevel">1</ac:parameter><ac:parameter ac:name="class">rm-contents</ac:parameter>
  <ac:parameter ac:name="exclude"></ac:parameter>
  <ac:parameter ac:name="type">list</ac:parameter><ac:parameter ac:name="outline">false</ac:parameter>
  <ac:parameter ac:name="include"></ac:parameter>
  </ac:structured-macro>
  <h2>Headings</h2>
  <p>The heading for this section is a level 2 heading</p>
  <h3>Level 3</h3>
  <p>Some text for level 3</p>
  <h4>Level 4</h4>
  <p>Some text for level 4</p>
  <h5>Level 5</h5>
  <p>Some text for level 5</p>
  <h2>Lists</h2>
  <h3>Unordered list</h3>
  <ul>
  <li>An</li>
  <li>unordered</li>
  <li>list</li>
  </ul>
  <h3>Ordered list</h3>
  <ol>
  <li>An</li>
  <li>Ordered</li>
  <li>List</li>
  </ol>
  <h3>Nested Unordered list</h3>
  <ul>
  <li>This list is nested</li>
  <li>This item has children<ul>
  <li>This is a child of item 2</li>
  <li>Another child of item 2</li>
  </ul>
  </li>
  <li>This is the third item</li>
  </ul>
  <h3>Nested Ordered list</h3>
  <ol>
  <li>This list is nested</li>
  <li>This item has children<ol>
  <li>This is a child of item 2</li>
  <li>Another child of item 2</li>
  </ol>
  </li>
  <li>This is the third item</li>
  </ol>
  <h2>Blocks</h2>
  <h3>Info Block</h3>
  <p><ac:structured-macro ac:name="info"><ac:rich-text-body><p>This should be an info block</p></ac:rich-text-body></ac:structured-macro></p>
  <h3>Note/Warning Block</h3>
  <p><ac:structured-macro ac:name="note"><ac:rich-text-body><p>This should be a note block</p></ac:rich-text-body></ac:structured-macro></p>
  <h3>Warning Block</h3>
  <p><ac:structured-macro ac:name="warning"><ac:rich-text-body><p>This should be a warning block</p></ac:rich-text-body></ac:structured-macro></p>
  <h2>Code</h2>
  <h3>No language</h3>
  <ac:structured-macro ac:name="code"><ac:parameter ac:name="theme">Midnight</ac:parameter><ac:parameter ac:name="linenumbers">true</ac:parameter><ac:parameter ac:name="language">none</ac:parameter><ac:plain-text-body><![CDATA[This is a code block with on language assigned
  ]]></ac:plain-text-body></ac:structured-macro>
  <h3>Javascript</h3>
  <ac:structured-macro ac:name="code"><ac:parameter ac:name="theme">Midnight</ac:parameter><ac:parameter ac:name="linenumbers">true</ac:parameter><ac:parameter ac:name="language">javascript</ac:parameter><ac:plain-text-body><![CDATA[var block = has_javascript_lang();
  ]]></ac:plain-text-body></ac:structured-macro>
  <h2>Tables</h2>
  <table>
  <thead>
  <tr>
  <th>Heading 1</th>
  <th>Heading 2</th>
  <th>Heading 3</th>
  </tr>
  </thead>
  <tbody>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  <tr>
  <td>Data 1</td>
  <td>Data 2</td>
  <td>Data 3</td>
  </tr>
  </tbody>
  </table>
  '''
# ---
# name: test_converter_image_test
  '''
  <h1>Page with Image</h1>
  <p><img alt="My Image" src="./images/image1.jpg" /></p>
  '''
# ---
# name: test_toc_convert
  '<h1>First heading</h1> <p><ac:structured-macro ac:name="toc" ac:schema-version="1"/></p> <h2>Second heading</h2>'
# ---
